### PR TITLE
Fix crash when NSUserData is corrupted

### PR DESCRIPTION
**Pull request by vvechkanov**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Fix crash when UserData is corrupted on device.

## Related PRs or issues

[AB#80422](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/80422)

### CODE REVIEW

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 1]
```diff
@@ -47,7 +47,12 @@ + (NSObject *)unarchiveKeyedData:(NSData *)data {
   } else {
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wdeprecated"
-    return [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    @try {
+      return [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    }
+    @catch(id) {
+      return nil;
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:54]: Add logs

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ Add support for Mac Catalyst to App Center Analytics and App Center Crashes. Als
 * **[Feature]** Add support for Mac Catalyst.
 * **[Fix]** Fix `Undefined symbol: ___llvm_profile_runtime` when the SDK was built from source by Carthage.
 * **[Fix]** Fix screen resolution detection macOS.
+* **[Fix]** Fix crash when NSData is corrupted.
```
Comment (MatkovIvan) [inline CHANGELOG.md:12]: What data? Please specify what data it is from customer's view. Class name of data structure says nothing

File: AppCenter/AppCenter/Internals/Util/MSUtility.m (hunk 1 above)
Comment (MatkovIvan) [reply AppCenter/AppCenter/Internals/Util/MSUtility.m]: Not resolved

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 3]
```diff
@@ -40,15 +40,23 @@ + (NSString *)sdkVersion {
...
 #pragma clang diagnostic ignored "-Wdeprecated"
-    return [NSKeyedUnarchiver unarchiveObjectWithData:data];
+      return [NSKeyedUnarchiver unarchiveObjectWithData:data];
 #pragma clang diagnostic pop
+    }
+  }
+  @catch(NSException *ex) {
+    
+    // Return nil if unarchiving fails
+    NSLog(@"Unarchiving NSData object %@ fails with this error: %@", data, ex.reason);
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:58]: @vvechkanov we do not use `NSLog` directly for SDK logs.

```suggestion
    // Return nil if unarchiving fails.
    MSLogError([MSAppCenter logTag], @"Unarchiving NSData failed with error: %@", ex.reason);
```

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 4]
```diff
@@ -40,15 +40,23 @@ + (NSString *)sdkVersion {
 }
 
 + (NSObject *)unarchiveKeyedData:(NSData *)data {
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:42]: There are usages of this function that wrapped in `@try`/`@catch` we need to clean it and check if result is `nil` instead.

**Comment by vvechkanov** [comment]
Rebase on the latest develop.

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 5]
```diff
@@ -44,15 +46,23 @@ + (NSString *)sdkVersion {
...
-  if (@available(iOS 11.0, macOS 10.13, watchOS 4.0, *)) {
-    NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingFromData:data error:nil];
-    unarchiver.requiresSecureCoding = NO;
-    return [unarchiver decodeTopLevelObjectForKey:NSKeyedArchiveRootObjectKey error:nil];
-  } else {
+  @try {
+    if (@available(iOS 11.0, macOS 10.13, watchOS 4.0, *)) {
+      NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingFromData:data error:nil];
```
Comment (annakocheshkova) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:51]: Shouldn't we pass an error as ref here?

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 6]
```diff
@@ -44,15 +46,23 @@ + (NSString *)sdkVersion {
...
+      unarchiver.requiresSecureCoding = NO;
+      return [unarchiver decodeTopLevelObjectForKey:NSKeyedArchiveRootObjectKey error:nil];
```
Comment (annakocheshkova) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:53]: And here

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 7]
```diff
@@ -44,27 +46,57 @@ + (NSString *)sdkVersion {
...
+      unarchivedData = [NSKeyedUnarchiver unarchiveObjectWithData:data];
 #pragma clang diagnostic pop
+    }
   }
+  @catch(NSException *ex) {
+    exception = ex;
+  }
+  if (!unarchivedData || exception) {
+    
+    //Unarchiving process failed
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:69]:
```suggestion
    // Unarchiving process failed.
```

File: AppCenter/AppCenter/Internals/Util/MSUtility.m [hunk 8]
```diff
@@ -44,27 +46,57 @@ + (NSString *)sdkVersion {
...
+      archivedData = [NSKeyedArchiver archivedDataWithRootObject:data];
 #pragma clang diagnostic pop
+    }
+  }
+  @catch(NSException *ex) {
+    exception = ex;
+  }
+  if (!archivedData || exception) {
+    
+    //Unarchiving process failed
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Util/MSUtility.m:95]:
```suggestion
    // Unarchiving process failed.
```

**Pull request merged**